tutorials/src/bin/basic-tutorial-9.rs

```diff
@@ -0,0 +1,184 @@
+extern crate failure;
+use failure::Error;
+
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_pbutils as gst_pbutils;
+use gst_pbutils::{
+    prelude::*, Discoverer, DiscovererContainerInfo, DiscovererInfo, DiscovererResult,
+    DiscovererStreamInfo,
+};
+
+use std::env;
+
+#[path = "../tutorials-common.rs"]
+mod tutorials_common;
+
+fn send_value_as_str(v: &glib::SendValue) -> Option<String> {
+    if let Some(s) = v.get::<&str>() {
+        Some(s.to_string())
+    } else if let Some(serialized) = v.serialize() {
+        Some(serialized)
+    } else {
+        None
+    }
+}
+
+fn print_stream_info(info: &DiscovererStreamInfo, depth: usize) {
+    let caps_str = if let Some(caps) = info.get_caps() {
+        if caps.is_fixed() {
+            gst_pbutils::pb_utils_get_codec_description(&caps)
+                .unwrap_or_else(|| String::from("unknown codec"))
+        } else {
+            caps.to_string()
+        }
+    } else {
+        String::new()
+    };
+
+    let stream_nick = info.get_stream_type_nick();
+    println!(
+        "{stream_nick:>indent$}: {caps_str}",
+        stream_nick = stream_nick,
+        indent = 2 * depth + stream_nick.len(),
+        caps_str = caps_str
+    );
+
+    if let Some(tags) = info.get_tags() {
+        println!("{:indent$}Tags:", " ", indent = 2 * depth);
+        for (tag, values) in tags.iter_generic() {
+            let mut tags_str = format!(
+                "{tag:>indent$}: ",
+                tag = tag,
+                indent = 2 * (2 + depth) + tag.len()
+            );
+            // reviewer: I'm intending to write something along the lines of
+            // println!("{}", values.iter().collect(send_value_as_str).join(",")
```

Comment:
That would allocate a temporary vector though :) https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.join might be an option here, or you can do it with `Iterator::fold()` but the way you do it is also fine

Reply:
For this code I was more interested in readability than performance, but since it's OK as it is, I'll leave it be :)